Readme.md
# What is CE-toy?

CE-toy is a toy implementation of a credit evaluation service. Credit evaluation is the business process that takes place when one or more applicants apply for a loan (or some other event that require credits). Data about the applicants are retrieved and evaluated according to a fixed framework called a credit process.

# Overall design

In CE-toy, a credit evaluation has been separated into two concerns:

* Loading of applicant data
* Evaluation of credit process rules

These topics are treated in separate sections below. One of the benefits of this separation is that the impure loading of data is kept separate from the pure domain logic related to evaluation of given credit process rules.

# Loading of a applicant data

Applicant data is often provided by 3rd party services. It is important to keep the requests to a minimum both from a cost perspective but, arguably more important, to avoid an excessive number of lookups for individual customers since such lookups may have a negative impact on their credit score.

A flora of 3rd party services exists and the datasets are heterogeneous often with overlaps and associated with different costs and different impacts on the customer credit scores. 

Based on this observation retrieving of applicant data can be viewed as an optimization problem.

> Given a set of known parameters and a set of required parameters we want to select a subset of all the available data loaders such that the cost is minimized while still all required parameters are provided.

It should also be noted that, in case of a rejected application, all the rules might not be evaluated and thus the actually required parameters can be a subset of all the parameters used by the credit process. This indicates that some kind of on demand loading can be useful.

## Modelling

Parameters in CE-toy are modelled as key-value pairs defined as

```haskell
type Key = String
data Value = IntValue Int | StringValue String
```

A data loader is then give by the following record

```haskell
data Loader = Loader { 
    name   :: String,    -- Some human friendly descriptive name
    cost   :: Int,       -- The cost for executing the load function 
    input  :: Set Key,   -- The required input parameters
    output :: Set Key,   -- The loaded output parameters
    load   :: Map Key Value -> IO (Map Key Value) -- The actual load function. The input dictionary is assumed to define all input keys and the output dictionary should define all output keys. The load function allows for arbitrary IO side effects.
  }
```

## Optimization

Since the number of data loaders are quite low an exhaustive search is feasible and it is probably also the simplest to implement.

The CE-toy does indeed implement this search strategy through the solve function. The solve function implements the following signature:

```
solve :: Set Key -> Set Key -> [Loader] -> [Loader]
solve knownKeys requiredKeys allLoaders = selectedLoaders
    where selectedLoaders = ...
```

The actual implementation has a slightly more general signature but, for the sake of reasoning, it does not matter.

# Evaluation of credit process rules

Credit rules might be intricate and to express them we need an expressive language. One option is to build up a new specialized language from scratch but that is both time consuming and hard. A common solution is instead to create a DSL (domain specific language) within the host programming language. That is the approach used here. A requirement from the data loading is that it must be possible to analyze the rules to find out which applicant data parameters that is used and thus might need to be loaded. One solution that satisfies these constraints is the [free monad interpreter pattern](https://softwareengineering.stackexchange.com/questions/242795/what-is-the-free-monad-interpreter-pattern). It might sound scary but using it in practice does not require a deep understanding of the underlying mechanism. Instead it can just be seen as a pattern to follow.

The proposed DSL consists of *rules* and *rule expressions*. Rules are basically named rule expressions that can be linked together in a chain. 

```haskell
data Rule = Rule String (RuleExpr Amount) | RAndThen Rule Rule
```

A rule expression must be able to express retrieval of data parameters and the requested credit amount. Data parameters might be of different types. String and int values are currently supported.

```haskell
data RuleExprF a = 
      GetIntValue Key (Int -> a) 
    | GetStringValue Key (String -> a)
    | GetAmount (Amount -> a)  
        deriving Functor

type RuleExpr = Free RuleExprF
```

The definition is a bit cryptic but lets ignore that and just accept it as a pattern. 

In order to get a proper DSL and not just data constructors the following helper functions are defined:

```haskell
rule :: String -> RuleExpr Amount -> Rule
rule name expr = Rule name expr

andThen :: Rule -> Rule -> Rule
andThen r1 r2 = RAndThen r1 r2

getIntValue :: Key -> RuleExpr Int
getIntValue key = liftF $ GetIntValue key id

getStringValue :: Key -> RuleExpr String
getStringValue key = liftF $ GetStringValue key id

getAmount :: RuleExpr Amount
getAmount = liftF $ GetAmount id
```

By adding the following trivial rules we can enhance readability.

```haskell
accept :: RuleExpr Amount
accept = getAmount

reject :: RuleExpr Amount
reject = return 0
```

We can now benefit from the trick used in the definition of the *rule expressions* and build up an entire credit process. It might look something like below.

```haskell
absoluteMaxAmount :: Amount -> RuleExpr Amount
absoluteMaxAmount amountLimit = do
    amount <- getAmount
    return $ min amount amountLimit
        
maxTotalDebt :: Int -> RuleExpr Amount
maxTotalDebt debtLimit = do
    creditA <- getIntValue "CreditA"
    creditB <- getIntValue "CreditB"
    if creditA + creditB > debtLimit then
        reject
    else
        accept

ceProcessExample :: Rule
ceProcessExample =
    rule "absoluteMaxAmount" (absoluteMaxAmount 1000) `andThen`
    rule "maxTotalDebt" (maxTotalDebt 500)

```

It might be worth taking a pause here just to contemplate what we got so far. We have an expressive language that has been bootstrapped from the host language. The rules can be expressed in a terse and clear manner without having to worry about implementation details of how it will be executed. Rules are also composable; a chain of rules is just another rule which simplifies reasoning. Moreover, the rules are entirely pure and free of side effects; in fact the rules are just a data structure.

## Evaluation of rules

Given the DSL for rules and rule expressions it is possible to construct data structures that represents a credit evaluation process. However, to actually evaluate it we need an interpreter.

A possible signature for a rule interpreter is

```haskell
ruleI :: Rule -> Amount -> Map Key Value -> [Loader] -> IO Amount
ruleI r = \requestedAmount knownParams allLoaders -> ...
```

In plain english this means

> **ruleI** is a function that takes a credit evaluation process, expressed as a Rule, and returns an evaluation function. The evaluation function, in turn, takes the requested amount, the known data parameters for the applicant and the available data loaders and it returns the granted amount.

Implementing the rule interpreter requires interpreting rule expressions. The rule expression, while simple, is a bit more technical. At its core is the implementation of retrieval of data parameters.

```haskell
getValue key = do
    kvs <- cache <$> get
    case Map.lookup key kvs of
        Just value -> return value
        Nothing -> do
            ls <- loaders <$> get
            case ls of
                [] -> throwError $ "could not produce a value for key " ++ key
                (l:ls') -> do
                    kvs' <- liftIO $ load l kvs          
                    modify (\s -> s { cache = Map.union kvs kvs', loaders = ls' })
                    getValue key
```

This function store a cache of all known parameters so far and a list of loaders. When a value is requested, the cache will first be queried. If the value exists in the cache the value is simply returned. If not, the first data loader in the list will be executed. The cache will be updated with the new values retrieved and getValue is executed again, recursively.

## Version control

It is important to be able to track changes in the credit evaluation process. Due to the fact that it is simply a file we can use ordinary version control tools like git to handle that.

# Putting the pieces together

Main.hs in CE-toy defines a number of loaders (all in-memory, trivial ones). The main method applies ruleI to an example credit process called ceProcessExample and retrieves an evaluation function. The evaluation function is then invoked with a sample amount, a national id and the list of loaders.

Executing main will produce the following output:

> result=(750,[("Rule evaluated absoluteMaxAmount.",1800,1000),("Rule evaluated maxAmountForAge.",1000,750),("Rule evaluated maxTotalDebt.",750,750),("Rule evaluated hasAddress.",750,750)])

# Limitations and future work

CE-toy only support single applicant scenarios. It seems possible to lift the implementation to handle multiple applicants using a few primitives but this has neither been well thought through nor implemented.

Currently the implementation is done in Haskell. Although Haskell is a good language for DSL embedding it is arguably esoteric. Investigate a C# solution. Have a look at [Sprache](https://github.com/sprache/Sprache). By leveraging linq and expression trees it might be possible to create a dsl where the following would be possible:

```csharp
class Process
{
    public static RuleExpr<int> AbsoluteMaxAmount(int amountLimit)
    {
        return
            from amount in Dsl.GetAmount()
            select Math.Min(amount, amountLimit);
    }

    public static RuleExpr<int> MaxTotalDebt(int debtLimit)
    {
        return
            from creditA in Dsl.GetValue("CreditA")
            from creditB in Dsl.GetValue("CreditB")
            from amount in Dsl.GetAmount()
            let totalCredit = creditA + creditB
            select totalCredit > debtLimit ? 0 : amount;
    }
}
```


CHANGELOG.md
# Revision history for ce-toy

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


Dsl.hs
{-# LANGUAGE DeriveFunctor #-}

module Dsl where

import Types
import Free 

data RuleExprF a = 
      GetIntValue Key (Int -> a) 
    | GetStringValue Key (String -> a)
    | GetAmount (Amount -> a)
    | SetAmount Amount a
    | IfB Bool (RuleExpr a) (RuleExpr a)
        deriving Functor

type RuleExpr = Free RuleExprF

getIntValue :: Key -> RuleExpr Int
getIntValue key = liftF $ GetIntValue key id

getStringValue :: Key -> RuleExpr String
getStringValue key = liftF $ GetStringValue key id

getAmount :: RuleExpr Amount
getAmount = liftF $ GetAmount id

setAmount :: Amount -> RuleExpr ()
setAmount amount = liftF $ SetAmount amount ()

ifB :: Bool -> RuleExpr a -> RuleExpr a -> RuleExpr a
ifB b t f = liftF $ IfB b t f

data Rule = Rule String (RuleExpr Amount) | RAndThen Rule Rule

rule :: String -> RuleExpr Amount -> Rule
rule name expr = Rule name expr

andThen :: Rule -> Rule -> Rule
andThen r1 r2 = RAndThen r1 r2

class IfThenElse b where
  ifThenElse :: b -> RuleExpr a -> RuleExpr a -> RuleExpr a

instance IfThenElse Bool where
  ifThenElse b t f = ifB b t f

accept :: RuleExpr Amount
accept = getAmount

reject :: RuleExpr Amount
reject = setAmount 0 >> getAmount

acceptLimit :: Amount -> RuleExpr Amount
acceptLimit limit = setAmount limit >> getAmount


Eval.hs
{-# LANGUAGE Rank2Types, TypeOperators, GeneralizedNewtypeDeriving, StandaloneDeriving #-}

module Eval where

import Data.Set (Set)
import qualified Data.Set as Set
import Data.Map (Map)
import qualified Data.Map as Map
import Control.Monad.State.Class
import Control.Monad.Except
import Control.Monad.State.Lazy
import Control.Monad.Writer.Lazy
import Types
import Loader
import Dsl
import Optimizer(solve)
import Free
import Debug.Trace

--type f ~> g = forall x . f x -> g x

data RuleExprEvalState = RuleExprEvalState {
    currentAmount :: Amount,
    cache :: Map Key Value,
    loaders :: [Loader]
  }

deriving instance Show RuleExprEvalState

newtype RuleExprEvalT a = RuleExprEvalT { 
    runRuleExpr :: ExceptT String (StateT RuleExprEvalState IO) a 
  } deriving (Functor, Applicative, Monad, MonadState RuleExprEvalState, MonadError String, MonadIO)

-- https://repl.it/@daniel_brannstrom/lazy-vars#main.hs
getValue :: Key -> RuleExprEvalT Value
getValue key = do
    kvs <- cache <$> get
    case Map.lookup key kvs of
        Just value -> return value
        Nothing -> do
        ls <- loaders <$> get
        case ls of
            [] -> throwError $ "could not produce a value for key " ++ key
            (l:ls') -> do
            kvs' <- liftIO $ load l kvs          
            modify (\s -> s { cache = Map.union kvs kvs', loaders = ls' })
            getValue key

getIntValue' :: Key -> RuleExprEvalT Int
getIntValue' key = do
    value <- getValue key
    case value of 
        IntValue intValue -> return intValue
        _ -> throwError $ "invalid type"

getStringValue' :: Key -> RuleExprEvalT String
getStringValue' key = do
    value <- getValue key
    case value of 
        StringValue strValue -> return strValue
        _ -> throwError $ "invalid type"

getAmount' :: RuleExprEvalT Amount
getAmount' = currentAmount <$> get

ruleExprI :: RuleExprF ~> RuleExprEvalT
ruleExprI (GetIntValue key next) = do
    iv <- getIntValue' key 
    return $ next iv
ruleExprI (GetStringValue key next) = do
    iv <- getStringValue' key 
    return $ next iv
ruleExprI (GetAmount next) = do
    amount <- getAmount'
    return $ next amount
ruleExprI (SetAmount amount next) = do
    modify (\s -> s { currentAmount = amount })
    return $ next
ruleExprI (IfB b tNext fNext) = 
    interp ruleExprI (if b then tNext else fNext)

-- assumption: process uses rebindable if so that both true and false branches can be analyzed
getRuleExprKeys' :: RuleExpr a -> [String]
getRuleExprKeys' (Pure a) = []
getRuleExprKeys' (Free f) = 
    case f of
        (GetIntValue key next) -> key : getRuleExprKeys' (next 0)
        (GetStringValue key next) -> key : getRuleExprKeys' (next "")
        (GetAmount next)       -> getRuleExprKeys' (next 0)
        (SetAmount _ next) -> getRuleExprKeys' next
        (IfB _ tNext fNext) -> getRuleExprKeys' tNext ++ getRuleExprKeys' fNext

getRuleKeys' :: Rule -> [String]
getRuleKeys' (Rule name expr) =
    let keys = getRuleExprKeys' expr in
    trace ("rule " ++ name ++ " has keys " ++ show keys) keys
getRuleKeys' (RAndThen r1 r2) = getRuleKeys' r1 ++ getRuleKeys' r2

ruleI :: Rule -> Amount -> Map Key Value -> [Loader] -> IO (Amount,[Decision])
ruleI r = \amount initKeyValues loaders -> do
    let initKeys = Set.fromList $ Map.keys initKeyValues
    let requiredKeys = Set.fromList $ getRuleKeys' r
    let loaders' = solve initKeys requiredKeys loaders
    let initState = RuleExprEvalState {
        currentAmount = amount,
        cache = initKeyValues,
        loaders = loaders'
    }
    (finalState,decisionLog) <- runWriterT $ evalR' r initState
    return (currentAmount finalState,decisionLog)

-- https://repl.it/@daniel_brannstrom/CredEvalTest#eval.hs
evalR' :: Rule -> RuleExprEvalState -> WriterT [Decision] IO RuleExprEvalState
evalR' (Rule name expr) state = do
    let expr' = Free.interp ruleExprI expr
    (errorOrAmount,state') <- lift $ runStateT (runExceptT $ runRuleExpr expr') state 
    case errorOrAmount of
        Left str -> do
            tell [ ("Failed to evaluate rule " ++ name ++ ". " ++ str, currentAmount state, 0) ]
            return $ state' { currentAmount = 0}
        Right amount' -> do 
            tell [ ("Rule evaluated " ++ name ++ ".", currentAmount state, amount') ]
            return $ state' { currentAmount = amount' }

evalR' (RAndThen r1 r2) state = evalR' r1 state >>= evalR' r2


Free.hs
{-# LANGUAGE InstanceSigs, Rank2Types, TypeOperators, DeriveFunctor #-}

module Free where

data Free f a = Free (f (Free f a)) | Pure a

instance Functor f => Functor (Free f) where
  fmap :: (a->b) -> Free f a -> Free f b 
  fmap f (Pure a) = Pure $ f a
  fmap f (Free g) = Free $ fmap (fmap f) g 

instance Functor f => Applicative (Free f) where
  pure :: a -> Free f a
  pure = Pure
  (<*>) :: Free f (a->b) -> Free f a -> Free f b
  (Pure f) <*> as = fmap f as
  (Free faf) <*> as = Free $ fmap (<*>as) faf

instance Functor f => Monad (Free f) where
  return :: a -> Free f a
  return = Pure
  (>>=) :: Free f a -> (a -> Free f b) -> Free f b
  (Pure a) >>= fm = fm a
  (Free faf) >>= fm = Free $ fmap (>>=fm) faf

type f ~> g = forall x . f x -> g x

freeM :: (Functor f, Functor g) => f ~> g -> Free f ~> Free g
freeM phi (Pure a) = Pure a
freeM phi (Free faf) = Free $ phi $ fmap (freeM phi) faf 

monad :: Monad m => Free m ~> m
monad (Pure a) = pure a
monad (Free maf) = do
  af <- maf
  monad af

interp :: (Functor f, Monad m) => f ~> m -> Free f ~> m
interp phi = monad . freeM phi

liftF :: Functor f => f ~> Free f
liftF f = Free $ fmap Pure f


Main.hs
module Main where

import Data.Map (Map)
import qualified Data.Map as Map
import Data.Set (Set)
import qualified Data.Set as Set
import Process(ceProcessExample)
import Loader(Loader(..))
import Eval(ruleI)
import Types

loaders :: [Loader]
loaders = [
    Loader { 
      name = "zero", cost = 0, 
      input = Set.fromList [ "NationalId" ], 
      output = Set.fromList [ "Age" ],
      load = (\_ -> return $ Map.fromList [("Age",IntValue 90)])
    },
    Loader { 
      name = "small", cost = 1, 
      input = Set.fromList [ "NationalId" ], 
      output = Set.fromList [ "Name", "Address" ],
      load = (\_ -> return $ Map.fromList [("Name",StringValue "Pelle"),("Address", StringValue "Storgatan 20")])
    },
    Loader {
      name = "medium", cost = 2, 
      input = Set.fromList [ "NationalId" ], 
      output = Set.fromList [ "Name", "Address", "Salary" ],
      load = (\_ -> return $ Map.fromList [("Name",StringValue "Pelle"),("Address",StringValue "Storgatan 20"),("Salary",IntValue 300)])
    },
    Loader {
      name = "large", cost = 3, 
      input = Set.fromList [ "NationalId" ], 
      output = Set.fromList [ "Name", "Address", "Salary", "NumReminders", "NumChildren", "CreditA", "CreditB" ],
      load = (\_ -> return $ Map.fromList [("Name",StringValue "Pelle"),("Address",StringValue "Storgatan 20"),("Salary",IntValue 300),("NumReminders",IntValue 1),("NumChildren",IntValue 2),("CreditA",IntValue 10),("CreditB", IntValue 0)])
    }, 
    Loader {
      name = "scoring", cost = 0, 
      input = Set.fromList [ "Salary", "NumChildren", "NumReminders" ], 
      output = Set.fromList [ "Scoring" ],
      load = (\_ -> return $ Map.fromList [("Scoring",IntValue 3)])
    }
  ]

main :: IO ()
main = do
    let f = ruleI ceProcessExample
    let amount = 1800
    let initKeyValues = Map.fromList [("NationalId", StringValue "29840113")]  -- could reach out for local cache for more data
    result <- f amount initKeyValues loaders
    putStrLn $ "result=" ++ show result


Optimizer.hs
{-# LANGUAGE MultiParamTypeClasses, FlexibleInstances, FunctionalDependencies #-}
-- https://repl.it/@daniel_brannstrom/dataloaders#main.hs

module Optimizer where

import Data.Function (on)
import Data.List (minimumBy, partition, tails, inits)

class Node n => Edge e n | e -> n where
  source :: e -> n
  sink :: e -> n
  transition :: n -> e -> n
  weight :: e -> Int

class Node n where
  (.==) :: n -> n -> Bool

-- https://hackage.haskell.org/package/utility-ht-0.0.15/docs/src/Data.List.HT.Private.html#removeEach
removeEach :: [a] -> [(a, [a])]
removeEach =
     map (\(ys, pivot, zs) -> (pivot,ys++zs)) . splitEverywhere
  where
    splitEverywhere :: [a] -> [([a], a, [a])]
    splitEverywhere xs =
      map
          (\(y, zs0) ->
            case zs0 of
                z:zs -> (y,z,zs)
                [] -> error "splitEverywhere: empty list")
          (init (zip (inits xs) (tails xs)))

genSolve :: Edge e n => n -> n -> [e] -> [(Int,[e])]
genSolve curNode goalNode edges =
  let
    (directEdges, indirectEdges) = partition (\e -> source e .== curNode) [ e | e <- edges, not (sink e .== curNode) ]
  in
    if goalNode .== curNode then
      pure (0,[])
    else do
      (e,es) <- removeEach directEdges 
      (w,es') <- genSolve (transition curNode e) goalNode (es ++ indirectEdges)
      return (w + weight e, e:es')

-- todo: minimumBy assumes there is a solution (will result in a runtime exception foldl on empty list otherwise)
solve :: Edge e n => n -> n -> [e] -> [e]
solve curNode goalNode edges = snd $ minimumBy (compare `on` fst) (genSolve curNode goalNode edges)

Process.hs
{-# LANGUAGE RebindableSyntax #-}

module Process where

import Prelude
import Types
import Dsl

conditionalLookupTest :: Amount -> RuleExpr Amount
conditionalLookupTest amountLimit = do
    age <- getIntValue "Age"
    if age < 20 then do
        amount <- getAmount
        return $ min amount amountLimit
    else do
        name <- getStringValue "Name"
        amount <- getAmount
        return $ amount + length name + 123

absoluteMaxAmount :: Amount -> RuleExpr Amount
absoluteMaxAmount amountLimit = do
    amount <- getAmount
    return $ min amount amountLimit
    
maxAmountForAge :: Amount -> Age -> RuleExpr Amount
maxAmountForAge amountLimit ageLimit = do
    age <- getIntValue "Age"
    amount <- getAmount
    if age > ageLimit && amount > amountLimit then
        acceptLimit amountLimit
    else
        accept
    
maxTotalDebt :: Int -> RuleExpr Amount
maxTotalDebt debtLimit = do
    creditA <- getIntValue "CreditA"
    creditB <- getIntValue "CreditB"
    if creditA + creditB > debtLimit then
        reject
    else
        accept

hasAddress :: RuleExpr Amount
hasAddress = do
    address <- getStringValue "Address"
    if address == "" then
        reject
    else
        accept

ceProcessExample :: Rule
ceProcessExample =
    rule "conditionalLookupTest" (conditionalLookupTest 1000) `andThen`
    rule "absoluteMaxAmount" (absoluteMaxAmount 1000) `andThen`
    rule "maxAmountForAge" (maxAmountForAge 750 85) `andThen`
    rule "maxTotalDebt" (maxTotalDebt 500) `andThen`
    rule "hasAddress" hasAddress   
